docs/specification/current.md

```diff
@@ -240,45 +240,66 @@ A prefetch token is a placeholder in a prefetch template that is *_replaced by i
+
+For read, the prefetch token FHIRPath is limited to *only* the name of the context element - no further navigation within the context element is permitted.  For example, `{{context.draftOrders.id}}` is not a valid prefetch token for a read because it attempts to access the id field of the medication field.  If expanding the prefetch token does not result in a simple Resource/id string, the prefetch is in error and the prefetch cannot be returned.
+
+##### search prefetch
+The cue that a prefetch is a search is that it starts with a resource name followed by a '?'.  Search prefetches can be more complex than reads - and not all CDS clients will support all potential forms of search.  Additional considerations include:
+
+1. Prefetch templates can drill deeper into the context, navigating to child elements.  For example `{{context.draftOrders.id}}` or event `{{context.draftOrders.ofType(MedicationRequest).medicationReference.reference}}` would be valid tokens.
+2. Prefetch templates can resolve to an empty collection.  If the template resolves to no elements, the prefetch result will be __null__, regardless of where within the search expression the prefetch token falls.
+3. Prefetch templates can resolve to a collection of more than one element.  If this occurs, the collection is returned as single comma-delimited concatenated string.  E.g. `{{context.draftOrders.ofType(MedicationRequest).medicationReference.reference}}` might resolve to `Medication/123,http%3a//someserver.com/Medication/456,Medication/789`
+4. Prefetch templates can resolve to complex objects.  This is handled as follows:
+* If a collection element is a simple type, the value is returned and any extensions or 'id' elements are ignored.  If no value is specified for the element, the element is ignored
+* If a collection element is a Coding or Identifier, the element is converted to [token](http://hl7.org/fhir/search.html#token) syntax with code/value followed by '|' followed by system.
+5. Content expanded from a prefetch token may require HTTP escaping to ensure a valid URL fragment
+6. When dealing with prefetch tokens that resolve to relative references, in some cases, the search needs the resolved token to be an id or collection of ids rather than references. In this circumstance, the FHIRPath can be followed with ":" and the name of the resource whose references should be filtered.  Only relative references for the specified resource type are returned.  For example `{{context.draftOrders.ofType(MedicationRequest).requester.reference:Practitioner}}` would resolve to the a list of ids who are requesters that are also Practitioners.  For legacy reasons, this can also be done with the userId element by inserting the resource name into the context name.  E.g. `{{context.userPatientRoleId}}`, `{{context.userPractionerId}}`, `{{context.userPractitionerRoleId}}`, and `{{context.userRelatedPersonId}}`.
```

Comment:
I don't agree with the description of legacy equivalents, or else I don't understand it correctly. Why would the user context for invoking a hook necessarily be equivalent to the person writing the draft orders? ( I can see that they might be the same and often would be... But that's not the same as definitional equivalence)

Reply:
The equivalence is that in `userId`, we provide a mechanism to take a userId context that would normally resolve to "Patient/123" or "Practitioner/456, etc. and instead to reference the context element as `userPatientId` and have it resolve to exactly "123", but only if the reference was a Patient.
The new mechanism of throwing `:Patient` on the end of the path is more generic.  It could work for userId (e.g. `context.userId:Patient`), but it can also work for any other path (e.g. `context.order:ServiceRequest` or `context.draftOrders.requester.reference:Practitioner`).  

The need that we solved as a one-off with userId turns out to be a general-purpose issue.  And it's an issue that can't be solved by just changing the name of a context variable.  The new approach will work for any path that corresponds to a 'Reference.reference' or to some other string that happens to be a relative reference.

Reply:
That all makes sense, in terms of the semantics of identifying a reference. The thing I was concerned about was the expectation (implication) that the person requesting a draft order would necessarily be the same as the user.

Reply:
Ok.  That wasn't an intended implication at all.  If you have suggestions on how to re-word to avoid that implication, I'm happy to adjust.

Reply:
Just remove the sentence station with

>For legacy reasons, this can also be done with the userId element by inserting the resource name

That capability is to find elsewhere, right? While it's analogous it's not quite the same thing as traversing a resource path in the manner you're documenting for the new capability.

Reply:
No - it's no longer documented elsewhere (or it shouldn't be - if I left it in, it was by accident).  The intention is to move all of the 'search' stuff into one place.

Reply:
Okay, then I'd document it separately; confusing to try to explain it here in the context of another syntax (and it's not like you can communicate "hook user" as part of navigating fhirpath on a context resource because there's no context variable representing the user in, just a set representing user IDs)

Reply:
The requirement is identical in both places.  And if we had adopted this new mechanism initially, there never would have *been* a userPatientId, etc. syntax.  That syntax is now redundant and unnecessary.  So it certainly seems to be appropriate to describe it in the same place as the more general syntax that's solving the same problem...

Reply:
Maybe I'm just missing what you have in mind. Can you explain what this would have looked like if we had introduced this new capability previously? What resource would we be drilling into with fhirpath expressions? An example might clarify for me.

Reply:
Using the new syntax, instead of `{{context.userPatientId}}`, you'd have `{{context.userId:Patient}}` - and there'd be no such thing as userPatientId.

Reply:
(Obviously, given that userPatientId is in use, I'm not proposing to eliminate it.  It would remain as a legacy syntax.)